### PR TITLE
[TG-1403] Correctly specialise classes with generic arrays or generic types 

### PR DESCRIPTION
**Pull request by thk123**
Depends on: #1555 :white_check_mark: 

Previous specialisation code only dealt with types that were directly `T`, this extends this to recursively check arrays (`T []`) and generic types (`Generic<T>`) which can be specialised.

### CODE REVIEW

File: unit/testing-utils/require_type.cpp [hunk 1]
```diff
@@ -289,3 +289,20 @@ require_type::require_java_non_generic_class(const typet &class_type)
 
   return java_class_type;
 }
+
+/// Verify a given type is a symbol type, optionally with a specific identifier
+/// \param type: The type to check
+/// \param identifier: The identifier the symbol type should have
+/// \return The cast version of the input type
+const symbol_typet &require_type::require_symbol(const typet &type,
```
Comment (NlightNFotis) [inline unit/testing-utils/require_type.cpp]: This function looks very useful. Nicely done.

File: unit/testing-utils/require_type.cpp [hunk 2]
```diff
@@ -289,3 +289,20 @@ require_type::require_java_non_generic_class(const typet &class_type)
...
+                                           const irep_idt &identifier)
+{
+  REQUIRE(type.id()==ID_symbol);
+  const symbol_typet &result=to_symbol_type(type);
+  if(identifier!="")
+  {
+    REQUIRE(result.get_identifier()==identifier);
+  }
+  return result;
+
```
Comment (NlightNFotis) [inline unit/testing-utils/require_type.cpp]: Nit: can this blank line be deleted?

**Review by NlightNFotis — APPROVED** [review]
This looks solid. Nicely done.

**Comment by thk123** [comment]
Linter errors are related to lambdas

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 3]
```diff
@@ -98,6 +80,103 @@ symbolt generate_java_generic_typet::operator()(
...
+    auto component_identifier = to_java_generic_parameter(parameter_type)
+                                  .type_variable()
+                                  .get_identifier();
+
+    optionalt<size_t> results =
+      java_generics_get_index_for_subtype(generic_class, component_identifier);
+
+    INVARIANT(results.has_value(), "generic component type not found");
+
+    if(results)
```
Comment (mgudemann) [inline src/java_bytecode/generate_java_generic_type.cpp]: isn't this tested in the above invariant using `.has_value()`?

**Review by mgudemann — APPROVED** [review]
LGTM, modulo the potential simplification

**Comment by thk123** [comment]
Thanks for reviews - comments and formatting addressed, will create a TG pointer bump then this should be good to go.

**Pull request merged**